addr_manager_ds.go

```
@@ -0,0 +1,316 @@
+	mgr.setAddrs(p, addrs, ttl, false)
+}
+
+func (mgr *DatastoreAddrManager) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration, add bool) {
+	for i := 0; i < dsWriteRetries; i++ {
+		// keys to add to the TTL manager
+		var keys []ds.Key
+		batch, err := mgr.ds.Batch()
+		if err != nil {
+			log.Error(err)
+			return
+		}
+
+		for _, addr := range addrs {
+			if addr == nil {
+				continue
+			}
+
+			key, err := peerAddressKey(&p, &addr)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			keys = append(keys, key)
+
+			if ttl <= 0 {
+				batch.Delete(key)
+				continue
+			}
+			has, err := mgr.ds.Has(key)
+			if err != nil || !has {
+				mgr.subsManager.BroadcastAddr(p, addr)
+			}
+
+			// Allows us to support AddAddr and SetAddr in one function
+			if !has || !add {
+				if err := batch.Put(key, addr.Bytes()); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+		if err := batch.Commit(); err != nil {
+			log.Errorf("failed to write addresses for peer %s: %s\n", p.Pretty(), err)
+			continue
+		}
+		mgr.ttlManager.setTTLs(keys, ttl)
+		return
+	}
+	log.Errorf("failed to avoid write conflict for peer %s after %d retries\n", p.Pretty(), dsWriteRetries)
+}
+
+// UpdateAddrs will update any addresses for a given peer and TTL combination to
+// have a new TTL.
+func (mgr *DatastoreAddrManager) UpdateAddrs(p peer.ID, oldTTL time.Duration, newTTL time.Duration) {
+	prefix := ds.NewKey(p.Pretty())
+	mgr.ttlManager.updateTTLs(prefix, oldTTL, newTTL)
+}
+
+// Addrs Returns all of the non-expired addresses for a given peer.
+func (mgr *DatastoreAddrManager) Addrs(p peer.ID) []ma.Multiaddr {
+	prefix := ds.NewKey(p.Pretty())
+	q := query.Query{Prefix: prefix.String()}
+	results, err := mgr.ds.Query(q)
+	if err != nil {
+		log.Error(err)
+		return []ma.Multiaddr{}
+	}
+
+	var addrs []ma.Multiaddr
+	for result := range results.Next() {
+		addrbytes := result.Value.([]byte)
+		addr, err := ma.NewMultiaddrBytes(addrbytes)
+		if err != nil {
```

ignoring errors makes people searching for bugs sad :(